### PR TITLE
Lab 3 - Letter frequency implementation

### PR DESCRIPTION
**Pull request by madhavansu**
Fixes #63 

#### Changes proposed in this PR:

- Write a function that returns a mapping of each letter to its frequency
- Write a function that returns a sorted slice of strings with elements

### CODE REVIEW

File: 03_letters/madhavansu/main_test.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	// When
+	main()
+
+	// Then
+	expected := strconv.Quote(`a:2
+b:1
+`)
+	actual := strconv.Quote(buf.String())
+	r.Equalf(expected, actual, "Unexpected output in main()")
+}
```
Comment (anzboi) [inline 03_letters/madhavansu/main_test.go:42]: Add newline after ending block.

File: 03_letters/madhavansu/main_test.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+	"github.com/stretchr/testify/require"
+)
+
+var tests = []struct {
+	s    string
+	lmap map[rune]int
+	ss   []string
+}{
+
```
Comment (anzboi) [inline 03_letters/madhavansu/main_test.go]: Get rid of this empty line

File: 03_letters/madhavansu/main_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import (
+	"bytes"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+var tests = []struct {
+	s    string
```
Comment (anzboi) [inline 03_letters/madhavansu/main_test.go]: Try to use more descriptive variables. 
`s` -> `input`, `in`,
`lmap` -> `frequencyMap`, `letterMap`
`ss` -> `output`, `sorted`

File: 03_letters/madhavansu/main_test.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	expected := strconv.Quote(`a:2
+b:1
+`)
+	actual := strconv.Quote(buf.String())
+	r.Equalf(expected, actual, "Unexpected output in main()")
+}
+func TestLettersFrequency(t *testing.T) {
+	r := require.New(t)
+	for _, v := range tests {
+		expectedMap := letters(v.s)
```
Comment (anzboi) [inline 03_letters/madhavansu/main_test.go]: This is not the expected output, this is the actual output as computed by `letters`.

File: 03_letters/madhavansu/main.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+var out io.Writer = os.Stdout
+
+func main() {
+	fmt.Fprintln(out, strings.Join(sortLetters(letters("aba")), "\n"))
+}
+
+func letters(s string) map[rune]int {
+	m := make(map[rune]int)
```
Comment (anzboi) [inline 03_letters/madhavansu/main.go]: You can have cleaner initialisation code with `m := map[rune]int{}`. This is more of a preference thing.

File: 03_letters/madhavansu/main_test.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	ss   []string
+}{
+
+	{"pwopoip@oiopoi@",
+		map[rune]int{'o': 5, 'p': 4, 'i': 3, '@': 2, 'w': 1},
+		[]string{"@:2", "i:3", "o:5", "p:4", "w:1"}},
+	{"‣£ê‣££~ 3 3 3",
+		map[rune]int{'3': 3, '£': 3, ' ': 3, '‣': 2, 'ê': 1, '~': 1},
+		[]string{" :3", "3:3", "~:1", "£:3", "ê:1", "‣:2"}},
+}
```
Comment (anzboi) [inline 03_letters/madhavansu/main_test.go:25]: Would like to see a test case against an empty input string `""`.

**Review by anzboi — COMMENTED** [review]
Code is logically well written, structured and mostly well tested. I have a few comments on formatting and naming.

File: 03_letters/madhavansu/main_test.go (hunk 1 above)
Comment (madhavansu) [reply 03_letters/madhavansu/main_test.go]: Updated.

File: 03_letters/madhavansu/main_test.go (hunk 2 above)
Comment (madhavansu) [reply 03_letters/madhavansu/main_test.go]: Updated.

File: 03_letters/madhavansu/main_test.go (hunk 1 above)
Comment (madhavansu) [reply 03_letters/madhavansu/main_test.go:42]: Updated.

File: 03_letters/madhavansu/main_test.go (hunk 3 above)
Comment (madhavansu) [reply 03_letters/madhavansu/main_test.go]: Updated.

File: 03_letters/madhavansu/main_test.go (hunk 4 above)
Comment (madhavansu) [reply 03_letters/madhavansu/main_test.go]: Updated.

File: 03_letters/madhavansu/main.go (hunk 5 above)
Comment (madhavansu) [reply 03_letters/madhavansu/main.go]: Updated.

File: 03_letters/madhavansu/main_test.go (hunk 6 above)
Comment (madhavansu) [reply 03_letters/madhavansu/main_test.go:25]: Updated.

**Review by anzboi — APPROVED** [review]
LGTM

**Pull request merged**